File: java/MinStack.java
// Lintcode
// 12. Min Stack
// Implement a stack with min() function, which will return the smallest number in the stack.

// It should support push, pop and min operation all in O(1) cost.

// Example
// push(1)
// pop()   // return 1
// push(2)
// push(3)
// min()   // return 2
// push(1)
// min()   // return 1

import java.util.Stack;

public class MinStack {
    private Stack<Integer> stack;
    private Stack<Integer> minStack;
    public MinStack() {
        // do initialize if necessary
        stack = new Stack<Integer>();
        minStack = new Stack<Integer>();
    }

    public void push(int number) {
        // write your code here
        stack.push(number);
        if(minStack.size() == 0 || minStack.peek() >= number) {
            minStack.push(number);
        }
    }

    public int pop() {
        // write your code here
        if(minStack.peek().equals(stack.peek())) {
            minStack.pop();
        }
        return stack.pop();
        
    }

    public int min() {
        // write your code here
        return minStack.peek();
    }
}

File: language-api/python/min_stack.py
class MinStack:
    
    def __init__(self):
        # do intialization if necessary
        self.stack = []
        self.minStack = []

    """
    @param: number: An integer
    @return: nothing
    """
    def push(self, number):
        # write your code here
        self.stack.append(number)
        if len(self.minStack) == 0 or self.minStack[-1] >= number:
            self.minStack.append(number)

    """
    @return: An integer
    """
    def pop(self):
        # write your code here
        if self.stack[-1] == self.minStack[-1]:
            self.minStack.pop()
        return self.stack.pop()

    """
    @return: An integer
    """
    def min(self):
        # write your code here
        return self.minStack[-1]
